### PR TITLE
[WIP] tc39/pattern-matching

### PR DESCRIPTION
**Pull request by vincentdchan**
This is a new PR based on #6761 , an implementation of pattern matching.

Proposal: [tc39/proposal-pattern-matching](https://github.com/tc39/proposal-pattern-matching) [@zkat](https://github.com/zkat)

Examples:

```javascript
const res = await fetch(jsonService)
const val = match (res) {
  {status: 200, headers: {'Content-Length': s}} => `size is ${s}`,
  {status: 404} => 'JSON not found',
  {status} if (status >= 400) => throw new RequestError(res)
}
```

```javascript
<Fetch url={API_URL}>{
  props => match (props) {
    {loading} => <Loading />,
    {error} => <Error error={error} />,
    {data} => <Page data={data} />
  }
}
</Fetch>
```

Progress:

- [x] Confirm the new syntax [@zkat](https://github.com/zkat)
- [x] Modify parser for the new syntax
- [x] Transpiling syntax sugars into [pattycake](https://github.com/zkat/pattycake) for experiencing
- [ ] Receive feedbacks and improve syntax
- [ ] Transpiling syntax into optimized inline code

### CODE REVIEW

**Comment by hzoo** [comment]
I think it would be cool to use the readme examples and tests from the proposal repo @zkat made anyway?

**Comment by xtuc** [comment]
btw no @babel-bot preview?

**Comment by vincentdchan** [comment]
We can try the the pattrycake edition of pattern matching in [https://babeljs.io/repl/build/7431/](https://babeljs.io/repl/build/7431/), enable the `stage-0` option and try it!  💐💐💐

File: packages/babylon/src/parser/expression.js [hunk 1]
```diff
@@ -417,7 +417,252 @@ export default class ExpressionParser extends LValParser {
...
+    this.state.inAsync = oldInAsync;
+
+    this.state.inParameters = oldInParameters;
+  }
+
+  parseMatchPattern(): N.MatchExpressionPattern | null {
+    let node;
+
+    if (this.match(tt.braceL)) {
+      return this.parseObjectMatchPattern();
```
Comment (xtuc) [inline packages/babylon/src/parser/expression.js]: I'm just wondering why:

```js
const val = match (res) {
  {'b': s} => true
}
```
```
repl: Unexpected token (4:3)

  2 | 
  3 | const val = match (res) {
> 4 |   {'b': s} => true
    |    ^
  5 | }
```

While: 
```js
const {'b': s} = true;
```

Works fine. From my understanding they are both using this `parseObjectMatchPattern`?

Comment (vincentdchan) [reply packages/babylon/src/parser/expression.js]: Yep, it should be right, it's just not supported currently.

Comment (xtuc) [reply packages/babylon/src/parser/expression.js]: At first glance I thought that your code already supported it but you are right. That `parseMatchProperty` is missing the quote handling.

**Comment by xtuc** [comment]
`regenerator` is required by the `async-to-generator` transform (if you don't have the good generator support) and Babel won't include it for you. The user needs to explicitly include it (with `@babel/polyfill` or whatever).

I would suggest that if `pattycake` is not too big we can include it by default: either by require'ing or injecting it as a helper directly.

**Comment by zkat** [comment]
it's pretty small, and I plan on making it smaller, so we could totally minify + inject it, specially at this early stage. That would be my preference right now.

**Comment by vincentdchan** [comment]
@zkat Fixed! see: [https://babeljs.io/repl/build/7482/](https://babeljs.io/repl/build/7482/)

File: packages/babel-generator/src/generators/expressions.js [hunk 2]
```diff
@@ -22,6 +22,64 @@ export function DoExpression(node: Object) {
...
+export function MatchClause(node: Object) {
+  this.print(node.pattern, node);
+  this.token("=>");
+  this.space();
+  this.print(node.body, node);
+  this.newline();
+}
+
+export function MatchGuard(node: Object) {
+  this.workd("if");
```
Comment (MiracleBlue) [inline packages/babel-generator/src/generators/expressions.js]: Seems to be a typo here, should this be `word` instead of `workd`?

**Comment by legodude17** [comment]
image
Where did the match guard go?

**Comment by vincentdchan** [comment]
@legodude17 Not implemented yet. The proposal is now changing often, just wait 😊

**Comment by legodude17** [comment]
Ah ok, thanks you clearing that up!

File: packages/babel-generator/src/generators/expressions.js (hunk 2 above)
Comment (vincentdchan) [reply packages/babel-generator/src/generators/expressions.js]: Yep, it's a mistake, thx

**Comment by quarryman** [comment]
@vincentdchan hello. 
Any estimates on this PR merge date?

**Comment by vincentdchan** [comment]
@quarryman I'm sorry to tell this implementation is outdate. The [proposal](https://github.com/tc39/proposal-pattern-matching) changes mostly. Pattern matching becomes a statement instead of expression currently. But it's not the final version.

**Comment by zkat** [comment]
I'm presenting the proposal for stage 1 this week, so keep your eyes peeled ✨

**Comment by hzoo** [comment]
nice! Justin and I are attending, so cya there

**Comment by quarryman** [comment]
@zkat 

> I'm presenting the proposal for stage 1 this week, so keep your eyes peeled

Any success with proposal? Will it be merged after presented or its just new review to be still discussed

**Comment by zkat** [comment]
It's stage 1 now!

There's a chunk of pending changes coming up after the discussion but there's progress, at least. I think you can expect some level of instability again as the proposal is prepared for stage 2 (I assume the presentation will be in September, since I'm probably missing the July meeting).

**Comment by shindakioku** [comment]
@vincentdchan what about implementation? :c

**Comment by JuanCaicedo** [comment]
Hi everyone! What's a good way to try this out, asides from in the [babel repl](https://babeljs.io/repl/build/7482/)? As in, if I wanted to try writing some code with this, would I build babel locally, or maybe require a plugin that does this? Excited to play with it! 😄

**Comment by vincentdchan** [comment]
@shindakioku Because there are too many changes in the new proposal, I am working on a new branch and reimplementing it now. I guess I will finish it in the new few weeks cuz you know, it's really a big feature.

**Comment by vincentdchan** [comment]
@JuanCaicedo The current PR is outdate, I am reimplementing it based on the new proposal. If you want to try this version, just checkout to my branch and compile the babel.

**Comment by shindakioku** [comment]
@vincentdchan yep, i understand. Thank you!

**Comment by JuanCaicedo** [comment]
@vincentdchan I'm playing around with this now 😀 Where's the best place to ask you questions that aren't directly about the proposal? If you want to open up issues on your fork, that could be a good place 👍

**Comment by GreenEyez86** [comment]
👌

**Comment by jhpratt** [comment]
@vincentdchan Any updates on progress? I see there haven't been any commits in quite a while — is it just waiting on something more from TC39?

**Comment by jhpratt** [comment]
Thanks! Was just wondering if that was abandoned or still a WIP (albeit
inactive). I definitely appreciate the work!

**Comment by vincentdchan** [comment]
@JuanCaicedo Great! I will announce an initial version ASAP, and you can help after that.

**Comment by vincentdchan** [comment]
Hey, guys, we are moving to a new PR #9010 for the implementation of case statement. Sorry for waiting for a long time. It's coming. 🚗🚗

**Pull request closed without merging**